timemachine/fe/single_topology_v3.py

```diff
@@ -0,0 +1,539 @@
+    improper_idxs = it.get_idxs()
+
+    return setup_dummy_interactions(
+        bond_idxs,
+        bond_params,
+        angle_idxs,
+        angle_params,
+        improper_params,
+        improper_idxs,
+        dummy_group,
+        root_anchor_atom,
+        nbor_anchor_atom,
+    )
+
+
+def setup_dummy_interactions(
+    bond_idxs,
+    bond_params,
+    angle_idxs,
+    angle_params,
+    improper_idxs,
+    improper_params,
+    dummy_group,
+    root_anchor_atom,
+    nbor_anchor_atom,
+):
+    """
+    Setup interactions involving atoms in a given dummy group. The following rules are applied:
+
+    1) We only allow for interactions within a dummy group, never between different dummy groups.
+    2) We form the augmented_dummy_group = dummy_group + [root anchor], and:
+        i) we only allow bond, angle, improper torsions terms to be turned on within a dummy group.
+        ii) we disable all nonbonded and proper torsions involving atoms in dummy groups
+    3) We can form a secondary augmented group using nbor_anchor_atom, but we only allow interactions
+        involving angles [i,j,k] where i in dummy_group, j == root_anchor_atom, and k == nbor_anchor_atom
+
+    Our motivation for this is 1) to ensure that the dummy interactions are factorizable, and that 2) the
+    dummy system is in an "enhanced" state so we can sample over torsional barriers etc. efficiently.
+
+    Parameters
+    ----------
+    bond_idxs: list of 2-tuples
+        Bond idxs
+
+    bond_params: list of 2-tuples
+        Force constants and bond lengths
+
+    angle_idxs: list of 3-tuples
+        Angle spanned by (i,j,k)
+
+    angle_params: list of 2-tuples
+        Force constants and equilibrium angles
+
+    improper_idxs: list of 4-tuples
+        Trefoil idxs for improper torsions
+
+    improper_params: list of 3-tuples
+        Force constant, phase, periods
+
+    dummy_group: set or list of int
+        Atoms to be decoupled
+
+    root_anchor_atom: int
+        A core atom we want to anchor our dummy_group to
+
+    nbor_anchor_atom: int
+        Another core atom connected to root_anchor_atom to build an angle restraint off of.
+
+    Returns
+    -------
+    (bonded_idxs, bonded_params)
+        Returns bonds, angles, and improper idxs and parameters.
+    """
+
+    # copy interactions that involve only root_anchor_atom
+    dummy_bond_idxs = []
+    dummy_bond_params = []
+    dummy_angle_idxs = []
+    dummy_angle_params = []
+    dummy_improper_idxs = []
+    dummy_improper_params = []
+
+    # dummy_group may be a set in certain cases, so sanity check.
+
+    assert len(dummy_group) == len(list(dummy_group))
+    dummy_group = list(dummy_group)
+    # dummy group and anchor
+
+    dga = dummy_group + [root_anchor_atom]
+    for idxs, params in zip(bond_idxs, bond_params):
+        if all([a in dga for a in idxs]):
+            dummy_bond_idxs.append(tuple([int(x) for x in idxs]))  # tuples are hashable etc.
```

Comment:
nit: redundant brackets
```suggestion
            dummy_bond_idxs.append(tuple(int(x) for x in idxs))  # tuples are hashable etc.
```